File: bin/tutorial8.py
import pafpython;

paf = pafpython.paf;


def Indentation(count) : 
	res = "";
	for i in range(count) : 
		res = res +  "\t";
	return res;


def ListPrimitiveMember(var, indent, recursive) :
	count = var._getMemberCount_()._;
	for i in range(count) :
		member = var._getMember_(i);
		print(Indentation(indent) + member._name_._ + " : " + member._category_._name_._);

def ListEnumMember(var, indent, recursive) : 
	count = var._getEnumeratorCount_()._;
	for i in range(count) :
		member = var._getEnumerator_(i);
		print(Indentation(indent) + member._name_._ + " : " + '%d' % member._value_._);

def ListClassMember(var, indent, recursive) :
	count = var._getBaseClassCount_()._;
	for i in range(count) :
		baseClass = var._getBaseClass_(i);
		print(Indentation(indent) + ":" + baseClass._name_._);

	count = var._getMemberCount_(False)._;
	for i in range(count) :
		member = var._getMember_(i, False);
		print(Indentation(indent) + member._name_._ + " : " + member._category_._name_._);
		if recursive :
			ListMember(member, indent + 1, recursive);


def ListNamespaceMember(var, indent, recursive) : 
	count = var._getMemberCount_()._;
	for i in range(count) :
		member = var._getMember_(i);
		print(Indentation(indent) + member._name_._ + " : " + member._category_._name_._);
		if recursive :
			ListMember(member, indent + 1, recursive);


tab = {}
tab[paf.pafcore.Category.name_space._] = ListNamespaceMember;
tab[paf.pafcore.Category.class_type._] = ListClassMember;
tab[paf.pafcore.Category.primitive_type._] = ListPrimitiveMember;
tab[paf.pafcore.Category.enum_type._] = ListEnumMember;


def ListMember(var, indent, recursive) :
	if var._category_._ in tab : 
		tab[var._category_._](var, indent, recursive);


ListMember(paf,0,1);

File: bin/tutorial8.lua
function Indentation(count)
	res = "";
	for i = 1, count do
		res = res .. "\t";
	end
	return res;
end

function ListPrimitiveMember(var, indent, recursive)
	count = var:_getMemberCount_();
	for i = 1, count do
		member = var:_getMember_(i - 1);
		print(Indentation(indent) .. member._name_ .. " : " .. member._category_._name_);
	end
end

function ListEnumMember(var, indent, recursive)
	count = var:_getEnumeratorCount_();
	for i = 1, count do
		member = var:_getEnumerator_(i - 1);
		print(Indentation(indent) .. member._name_ .. " : " .. member._value_);
	end
end

function ListClassMember(var, indent, recursive)
	count = var:_getBaseClassCount_();
	for i = 1, count do
		baseClass = var:_getBaseClass_(i - 1);
		print(Indentation(indent) .. baseClass._name_);
	end
	count = var:_getMemberCount_(false);
	for i = 1, count do
		member = var:_getMember_(i - 1, false);
		print(Indentation(indent) .. member._name_ .. " : " .. member._category_._name_);
		if recursive then
			ListMember(member, indent + 1, recursive);
		end
	end
end

function ListNamespaceMember(var, indent, recursive)
	count = var:_getMemberCount_();
	for i = 1, count do
		member = var:_getMember_(i - 1);
		print(Indentation(indent) .. member._name_ .. " : " .. member._category_._name_);
		if recursive then
			ListMember(member, indent + 1, recursive);
		end
	end
end;

tab = {}
tab[paf.pafcore.Category.name_space._] = ListNamespaceMember;
tab[paf.pafcore.Category.class_type._] = ListClassMember;
tab[paf.pafcore.Category.primitive_type._] = ListPrimitiveMember;
tab[paf.pafcore.Category.enum_type._] = ListEnumMember;

function ListMember(var, indent, recursive)
	func = tab[var._category_._]
	if func then
		func(var, indent, recursive);
	end
end

function ListDerivedClass(var)
	it = var:_getFirstDerivedClass_();
	while (it)
	do
		cls = it:deref();
		print(cls._name_);
		it = it:next();
	end
end

--print(paf.pafcore.Category.name_space._name_);
ListMember(paf,0,0);
--ListDerivedClass(paf.pafcore.Metadata);